File: src/com/ptby/dynamicreturntypeplugin/json/ConfigAnalyser.java
package com.ptby.dynamicreturntypeplugin.json;

import com.intellij.openapi.project.Project;
import com.ptby.dynamicreturntypeplugin.config.ClassMethodConfigKt;
import com.ptby.dynamicreturntypeplugin.config.DynamicReturnTypeConfig;
import com.ptby.dynamicreturntypeplugin.config.FunctionCallConfigKt;
import com.ptby.dynamicreturntypeplugin.config.ProjectDynamicReturnTypeMap;
import com.ptby.dynamicreturntypeplugin.config.multi.OpenProjects;
import com.ptby.dynamicreturntypeplugin.config.multi.ProjectsConfigRefresher;
import com.ptby.dynamicreturntypeplugin.config.multi.RefreshProjectCallBack;

import java.util.List;

public class ConfigAnalyser implements ProjectConfigChangeListener, RefreshProjectCallBack {
    private final OpenProjects openProjects;
    private final ProjectDynamicReturnTypeMap projectDynamicReturnTypeMap = new ProjectDynamicReturnTypeMap();

    private final ProjectsConfigRefresher openProjectsConfigRefresher = new ProjectsConfigRefresher();

    public ConfigAnalyser( OpenProjects openProjects ) {
        this.openProjects = openProjects;
    }


    public DynamicReturnTypeConfig getCurrentConfig( Project project ) {
        return projectDynamicReturnTypeMap.get( project );
    }


    public List<ClassMethodConfigKt> getCurrentClassMethodConfigs( Project project ) {
        return projectDynamicReturnTypeMap.get( project ).getClassMethodConfigs();
    }


    public List<FunctionCallConfigKt> getCurrentFunctionCallConfigs( Project project ) {
        return projectDynamicReturnTypeMap.get( project ).getFunctionCallConfigs();
    }


    public void notifyProjectIsClosed( Project project ) {
        projectDynamicReturnTypeMap.resetProject( project );
    }


    public void setConfig( Project project, DynamicReturnTypeConfig config ) {
        projectDynamicReturnTypeMap.put( project , config );
    }

    @Override
    public void refreshAllConfigs() {
        Project[] openProjectsAsArray = openProjects.getOpenProjectsAsArray();
        openProjectsConfigRefresher.refreshProjects( openProjectsAsArray, this );
    }
}

File: src/com/ptby/dynamicreturntypeplugin/json/ConfigAnalyser.kt
package com.ptby.dynamicreturntypeplugin.json

import com.intellij.openapi.project.Project
import com.ptby.dynamicreturntypeplugin.config.ClassMethodConfigKt
import com.ptby.dynamicreturntypeplugin.config.DynamicReturnTypeConfig
import com.ptby.dynamicreturntypeplugin.config.FunctionCallConfigKt
import com.ptby.dynamicreturntypeplugin.config.ProjectDynamicReturnTypeMap
import com.ptby.dynamicreturntypeplugin.config.multi.OpenProjects
import com.ptby.dynamicreturntypeplugin.config.multi.ProjectsConfigRefresher
import com.ptby.dynamicreturntypeplugin.config.multi.RefreshProjectCallBack

class ConfigAnalyser(private val openProjects: OpenProjects) : ProjectConfigChangeListener,
                                                                      RefreshProjectCallBack {
    private val projectDynamicReturnTypeMap = ProjectDynamicReturnTypeMap()

    private val openProjectsConfigRefresher = ProjectsConfigRefresher()


    fun getCurrentConfig(project: Project): DynamicReturnTypeConfig {
        return projectDynamicReturnTypeMap.get(project)
    }


    fun hasArrayAccessEnabled(project: Project): Boolean {
        return getCurrentConfig(project).hasArrayAccessEnabled()
    }


    fun getCurrentClassMethodConfigs(project: Project): List<ClassMethodConfigKt> {
        return projectDynamicReturnTypeMap.get(project).classMethodConfigs
    }


    fun getCurrentFunctionCallConfigs(project: Project): List<FunctionCallConfigKt> {
        return projectDynamicReturnTypeMap.get(project).functionCallConfigs
    }


    fun notifyProjectIsClosed(project: Project) {
        projectDynamicReturnTypeMap.resetProject(project)
    }


    override fun setConfig(project: Project, config: DynamicReturnTypeConfig) {
        projectDynamicReturnTypeMap.put(project, config)
    }

    override fun refreshAllConfigs() {
        val openProjectsAsArray = openProjects.getOpenProjectsAsArray()
        openProjectsConfigRefresher.refreshProjects(openProjectsAsArray, this)
    }
}
